vault/src/system.rs

```diff
@@ -405,7 +409,14 @@ async fn run_and_monitor_tasks(shutdown_tx: ShutdownSender, items: Vec<(&str, Se
         publish_tokio_metrics(metrics_iterators),
     ));
 
-    let _ = join(tokio_metrics, join_all(tasks)).await;
+    match join(tokio_metrics, join_all(tasks)).await {
+        (Ok(Err(err)), _) => Err(err),
+        (_, results) => results
+            .into_iter()
+            .find(|res| matches!(res, Ok(Err(_))))
+            .and_then(|res| res.ok())
+            .unwrap_or(Ok(())),
```

Comment:
Nit: using try_join_all would clean this up a bit

Reply:
We can't because `try_join_all` would cancel all futures once the first error is encountered, bypassing the shutdown logic.

Reply:
Why would that be a problem?

Reply:
It's probably not a problem right now, IIRC there is no "shutdown logic" if we read the signal but we may have tasks dependent on it in the future. That being said I also don't usually like to maintain code for some expected future purpose but I don't want to figure out if that change would work right now - need to merge for audit.